File: 02_linked_lists/10_remove_node/remove_node.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

const removeNode = (head, targetVal) => {
  let current = head;
  let prev = null;

  while(current != null){
    if(current.val === targetVal && prev === null) {
      head = head.next;
      break;
    };
    if(current.val === targetVal && prev != null) {
      prev.next = current.next;
      break;
    };
    prev = current;
    current = current.next;
  }

  return head
};

module.exports = {
  removeNode,
};

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");

a.next = b;
b.next = c;
c.next = d;
d.next = e;
e.next = f;

// a -> b -> c -> d -> e -> f

console.log(removeNode(a, "c"));
// a -> b -> d -> e -> f


const x = new Node("x");
const y = new Node("y");
const z = new Node("z");

x.next = y;
y.next = z;

// x -> y -> z

console.log(removeNode(x, "z"));
// x -> y

const q = new Node("q");
const r = new Node("r");
const s = new Node("s");

q.next = r;
r.next = s;

// q -> r -> s

console.log(removeNode(q, "q"));
// r -> s

File: 02_linked_lists/10_remove_node/remove_node.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None

def remove_node(head, target_val):
  if head.val == target_val:
    return head.next;

  prev = None
  current = head

  while current is not None:
    if current.val == target_val:
      prev.next = current.next
      break
    prev = current
    current = current.next

  return head

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")
e = Node("e")
f = Node("f")

a.next = b
b.next = c
c.next = d
d.next = e
e.next = f

# a -> b -> c -> d -> e -> f

print(remove_node(a, "c"))

q = Node("q")
r = Node("r")
s = Node("s")

q.next = r
r.next = s

# q -> r -> s

print(remove_node(q, "q"))